TypeSystem/Upper-lower-bounds/up_A01_t01.dart

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion We define the upper bound of two types T1 and T2 to be UP(T1,T2)
+/// as follows.
+/// - UP(X1 & B1, T2) =
+///   - T2 if X1 <: T2
+///   - otherwise X1 if T2 <: X1
+///   - otherwise UP(B1a, T2) where B1a is the greatest closure of B1 with
+///     respect to X1, as defined in inference.md.
+/// - UP(T1, X2 & B2) =
+///   - X2 if T1 <: X2
+///   - otherwise T1 if X2 <: T1
+///   - otherwise UP(T1, B2a) where B2a is the greatest closure of B2 with
+///     respect to X2, as defined in inference.md.
+///
+/// @description Check that upper bound of intersection types is calculated as
+/// expected
+/// @author sgrekhov22@gmail.com
+
+int? foo() => DateTime.now().millisecondsSinceEpoch.isEven? null : 1;
+
+void f<X extends num>(bool b, X x) {
+  if (x is int) { // `x` promoted to `X & int`.
+    var y1 = b ? x : null;
+    y1.isEven; // Error. Type of `y1` is `int?`
+//     ^^^^^^
+// [analyzer] unspecified
+// [cfe] unspecified
+
+    var y2 = b ? foo() : x;
+    y2.isEven;
+//     ^^^^^^
+// [analyzer] unspecified
+// [cfe] unspecified
```

Comment:
Here we have `UP(int?, X & int) == UP(int?, int) == UP(int, int)? == int?`, which can be tested using `expectStaticType`.

The old type was `UP(int?, X & int) == UP(int, X & int)? == UP(int, int)? == int?`. This means that we get the same result for a different reason , and it makes a lot of sense to check that we still get `int?`. `expectStaticType` could do that, `X? w =x;` would confirm that the result isn't of the form `X & ...`.

Reply:
Thanks!